src/layer_lifecycle.rs

```diff
@@ -0,0 +1,362 @@
+    /// implementors to recover from this case by, for example, implementing migrations from older
+    /// versions of the metadata to the current one.
+    ///
+    /// The default implementation returns [`InvalidMetadataRecoveryResult::DeleteLayer`] to signal
+    /// that the existing layer should be deleted in its entirety.
+    fn recover_from_invalid_metadata(
+        &self,
+        #[allow(unused_variables)] layer_metadata: &toml::value::Table,
+        #[allow(unused_variables)] build_context: &BuildContext<P, BM>,
+    ) -> Result<MetadataRecoveryStrategy<LM>, E> {
+        // Default implementation is to delete the layer if the metadata is invalid
+        Ok(MetadataRecoveryStrategy::DeleteLayer)
+    }
+
+    /// Based on the current state of the layer, determines how the layer will be processed
+    ///
+    /// This will be called by [`execute_layer_lifecycle`] when the layer for this lifecycle already
+    /// exists. Based on the layer contents and its metadata, implementors can decide how the layer
+    /// will be processed.
+    ///
+    /// A common case is the comparison of the layer metadata to some expected value to determine
+    /// if the contents are still valid for the current build. If the contents are no longer valid,
+    /// an implementation might return [`ValidateResult::RecreateLayer`] which will result in the
+    /// layer being deleted and then recreated via [`LayerLifecycle::create`]. If the layer contents
+    /// are still valid, [`ValidateResult::KeepLayer`] could be returned to keep the layer as-is
+    /// without doing any modifications to it.
+    ///
+    /// The common case above is just one conceivable implementation of this function, take a look
+    /// at [`ValidateResult`] for additional validation results and their meanings.
+    fn validate(
+        &self,
+        #[allow(unused_variables)] layer_path: &Path,
+        #[allow(unused_variables)] layer_content_metadata: &LayerContentMetadata<LM>,
+        #[allow(unused_variables)] build_context: &BuildContext<P, BM>,
+    ) -> ValidateResult {
+        // Default implementation is to always recreate the layer
+        ValidateResult::RecreateLayer
+    }
+
+    /// Updates an existing layer
+    fn update(
+        &self,
+        #[allow(unused_variables)] layer_path: &Path,
+        #[allow(unused_variables)] layer_content_metadata: LayerContentMetadata<LM>,
+        #[allow(unused_variables)] build_context: &BuildContext<P, BM>,
+    ) -> Result<LayerContentMetadata<LM>, E> {
+        // Default implementation is a no-op
+        Ok(layer_content_metadata)
+    }
+
+    fn layer_lifecycle_data(
+        &self,
+        #[allow(unused_variables)] layer_path: &Path,
+        #[allow(unused_variables)] layer_content_metadata: LayerContentMetadata<LM>,
+    ) -> Result<O, E> {
+        Ok(O::default())
+    }
+
+    fn on_lifecycle_start(&self) {}
+    fn on_keep(&self) {}
+    fn on_update(&self) {}
+    fn on_create(&self) {}
```

Comment:
The `on_update` and `on_create` hooks here seem like they will always be called at the same time as `create` and `update`, thereby duplicating them. I'm presuming the reason these exist is because `on_keep` doesn't have an equivalent lifecycle method, and it would be inconsistent to just have `on_keep` and not the others? Or am I missing additional use-cases why `on_update` and `on_create` exist? :-)

Reply:
@Malax ^